### ISSUE TITLE
A Cancellable invocable with payload?

### ISSUE THREAD

**Issue by andremachado** [jamesmh/coravel#182]
Hello!

Would be nice to be able to have an invocable that's at the same time cancellable and that also contains a payload. What about QueueCancellableInvocableWithPayload()?

Thanks,

André

**Comment by tolbxela** [comment]
+1 to QueueCancellableInvocableWithPayload

You can achieve it also right now with a custom class.

Here is my base class with IInvocable, ICancellableTask and IInvocableWithPayload

```
 public class InvocableBase<T> : ICancellableTask, IInvocable, IInvocableWithPayload<T> {}
```

**Comment by mrb0nj** [comment]
@tolbxela I don't suppose you have a slightly more complete example do you? I can't see to get my head around where I need to call QueueInvocableWithPayload/QueueCancellableInvocable and where do i pass in the payload?!

Apologies in advance if i'm being extra stupid today!

**Comment by tolbxela** [comment]
### The base class:

```csharp
public class InvocableBase<T> : ICancellableTask, IInvocable, IInvocableWithPayload<T>
{
        public CancellationToken Token { get; set; }

        public T Payload { get; set; }

        public virtual Task Invoke()
        {
            return Task.CompletedTask;
        }
}
```
### The task:

```csharp
public class InvocableTask : InvocableBase<string>
{
        public override async Task Invoke()
        {
              // code
              System.Console.WriteLine(Payload);
        }
}
```
### Using of the InvocableTask:

```csharp

var TaskGuid = Queue.QueueInvocableWithPayload<InvocableTask , string>("Test");

```

**Comment by mrb0nj** [comment]
@tolbxela Thank you for the example, just one question though - and apologies if i'm missing the point massively - where does the CancellationTokenSource instance come from so i can cancel the task later on? Thanks in advance.

**Comment by tolbxela** [comment]
@mrb0nj: I guess it comes from here: 

https://github.com/jamesmh/coravel/blob/1df126935a7cc6a72ae879794edc3aa9164ae2da/Src/Coravel/Queuing/Queue.cs#L59

**Comment by lengockyquang** [comment]
any update ?

**Comment by lengockyquang** [comment]
> ### The base class:
> ```cs
> public class InvocableBase<T> : ICancellableTask, IInvocable, IInvocableWithPayload<T>
> {
>         public CancellationToken Token { get; set; }
> 
>         public T Payload { get; set; }
> 
>         public virtual Task Invoke()
>         {
>             return Task.CompletedTask;
>         }
> }
> ```
> 
> ### The task:
> ```cs
> public class InvocableTask : InvocableBase<string>
> {
>         public override async Task Invoke()
>         {
>               // code
>               System.Console.WriteLine(Payload);
>         }
> }
> ```
> 
> ### Using of the InvocableTask:
> ```cs
> var TaskGuid = Queue.QueueInvocableWithPayload<InvocableTask , string>("Test");
> ```

Thanks with your advice, but with your implementation, how can i cancel a specific task that already executed and in processing state

**Comment by tolbxela** [comment]
Here is the example in Docs: https://docs.coravel.net/Queuing/#queuing-cancellable-invocables

**Comment by jamesmh** [comment]
FYI, anyone wanting this - I have some relevant comments in #209 around the mechanics and public interface for this.

TLDR; -> eventually I want to remove `QueueCancellableInvocable` and just make Coravel automatically "know" based on whether the interface is present on the invocable or not whether a token should be generated for the task. A new method on the queue like `TryCancel()` would be available instead of returning the token to the caller. See comment on #209. Open to comments etc. 🙂

**Comment by jamesmh** [comment]
As per the comment above: 

- Returning the token is a leaky abstraction
-  There are some potential existing concurrency issues that arise from that method too
- There's a loss of modularity in having dedicated methods like `QueueCancellableInvocableWithPayload` - what if we need that functionality PLUS some other new behaviors? The method names become crazy long, hard to sort through and very rigid.

**Comment by snargledorf** [comment]
Couldn't an overload of QueueInvocableWithPayload be created that accepts a cancellation token? This way the caller can control cancellation using the standard cancellation pattern.

**Comment by jamesmh** [comment]
> Couldn't an overload of QueueInvocableWithPayload be created that accepts a cancellation token? This way the caller can control cancellation using the standard cancellation pattern.

Since this method comes from an interface, the invocables would have to implement 2 methods now. The caller would not know which method has the correct logic filled/coded out.

The signature of the method could be changed to have an optional/nullable cancellation token param. Considerations that comes up from this might include"
- Now the caller would have to create a token for every single invocable instance - whether or not it's used. 
- Do we still use the interface that exists and cause a breaking change?
- Do we keep both methods? Would seem odd...

If this option were taken (under consideration) then it would 100% lead to removing the use of the interface to mark the usage of a cancellation token given the points above. 

There's a question around philosophy here too: do the ".NET way" of doing things here or "something else that represents 'the coravel way' instead?"

I think there were other choices made that are worth changing the let's say a more ".NET way" of doing things, but I also think there are some decisions were Coravel deviates a bit from the norm (which is kinda what sparked Coravel in the first place years ago 😅).

For now, all things worth considering by any one whos participating in this or other discussions.

Would love to hear what others think on this - does anyone have the same thoughts as before? Any changed thoughts? Indifferent? Something else? 🙂

**Comment by brainiac256** [comment]
Would it make sense to, instead of returning the token on enqueuing, expose a method on the queue interface `IQueue.GetCancellationTokenSourceFor(Guid taskGuid)`?

Then the workflow would look like:

```C#
// in some controller, the invocable is queued
Guid taskGuid = _queue.QueueInvocable<MyInvocable>();
// or 
Guid taskGuid = _queue.QueueInvocableWithPayload<MyParameterizedInvocable>(payload);
// coravel code detects whether the `MyInvocable` is a `ICancellableTask` and generations a cancellation token source or not, doesn't matter to the app code right now
return OK(taskGuid);
// taskGuid is stored in the app somewhere for the user to come back to later
// ...
// five minutes later, the user decides the task has taken too long and wants to cancel it
CancellationTokenSource cts = _queue.GetCancellationTokenSourceFor(taskGuid);
cts.Cancel();
```

Since you're already storing the cancellation token source objects in a `Dictionary<Guid, CancellationTokenSource>` this would A) be performant and B) keep the application developer from having to essentially duplicate that dict in their application code.

The new method could throw something like `NotImplementedException` or something more specific if the app code calls it on a task that wasn't registered as an `ICancellableTask`, or just implement it as a `bool TryGetCancellationTokenSource(Guid taskGuid, out CancellationTokenSource cts)` that basically passes through to the dictionary.

A special case would probably be enqueuing anonymous async tasks... maybe overload `QueueAsyncTask` with `QueueAsyncTask(Func<CancellationToken, Task>)`? That would be consistent with other async invokes in the dotnet ecosystem, I think.

```C#
_queue.QueueAsyncTask( async(CancellationToken tk) =>
{
  //...
  await OtherCancellableAsyncOperation(tk);
  //...
});
```
